src/core/exec-invoke.c

```diff
@@ -2251,6 +2252,128 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
         return 1;
 }
 
+static int can_mount_proc(const ExecContext *c, ExecParameters *p) {
+        _cleanup_close_pair_ int errno_pipe[2] = EBADF_PAIR;
+        _cleanup_(sigkill_waitp) pid_t pid = 0;
+        _cleanup_(rmdir_and_freep) char *tmpdir = NULL;
+        ssize_t n;
+        int r;
+
+        assert(c);
+        assert(p);
+
+        /* If running via unprivileged user manager and /proc is masked (e.g. /proc/kmsg is over-mounted with tmpfs
+         * like systemd-nspawn does), then mounting /proc will fail with EACCES. This is due to a kernel restriction
+         * where unprivileged user namespaces cannot mount a less restrictive instance of /proc. Since this restriction
+         * only applies to user managers, we skip this check for other scopes.
+         */
+        if (p->runtime_scope != RUNTIME_SCOPE_USER)
+                return 1;
```

Comment:
that's not strictly true. in theory we do support systems without CAP_SYS_ADMIN (i.e. in a container world, where the container manager sets up *all* mounts, so that we don't have to anymore).

maybe do return have_effective_capability(CAP_SYS_ADMIN)?

Reply:
Odd this isn't working as expected. `have_effective_cap(CAP_SYS_ADMIN)` is returning true for unprivileged user managers but fails to mount /proc. Hence tests are failing. Let me debug further

Reply:
Okay another fun fact (these are getting less fun as time goes on): the kernel checks that the process has CAP_SYS_ADMIN within the user namespace corresponding to the mount namespace. So we actually need to do this check in the temporary process.

Reply:
Fixed by removing this check altogether

Reply:
Changed this to check CAP_SYS_ADMIN prior to entering userns to avoid PID1 re-checking if it can mount /proc over and over. Seems like a waste